ext/standard/streamsfuncs.c

```diff
@@ -756,116 +809,141 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
-		set_count = stream_array_to_fd_set(r_array, &rfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = -1;
+    int retval = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
+
+    // Determine the largest file descriptor
+    if (r_array != NULL) {
+        max_fd = stream_array_get_max_fd(r_array);
+    }
+    if (w_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(w_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
+    if (e_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(e_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
 
-	if (w_array != NULL) {
-		set_count = stream_array_to_fd_set(w_array, &wfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    if (max_fd == -1) {
+        zend_value_error("No stream arrays were passed or all arrays were empty");
+        RETURN_THROWS();
+    }
 
-	if (e_array != NULL) {
-		set_count = stream_array_to_fd_set(e_array, &efds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    // Allocate fd_bigset based on max_fd
+    FD_BIGSET_ZERO(&rfds, max_fd + 1);
+    FD_BIGSET_ZERO(&wfds, max_fd + 1);
+    FD_BIGSET_ZERO(&efds, max_fd + 1);
 
-	if (!sets) {
-		zend_value_error("No stream arrays were passed");
-		RETURN_THROWS();
-	}
+    // Populate the fd_bigset structures
+    if (r_array != NULL) {
+        stream_array_to_fd_bigset(r_array, &rfds, &max_fd);
+    }
 
-	if (!PHP_SAFE_MAX_FD(max_fd, max_set_count)) {
-		RETURN_FALSE;
-	}
+    if (w_array != NULL) {
+        stream_array_to_fd_bigset(w_array, &wfds, &max_fd);
+    }
 
-	if (secnull && !usecnull) {
-		if (usec != 0) {
-			zend_argument_value_error(5, "must be null when argument #4 ($seconds) is null");
-			RETURN_THROWS();
-		}
-	}
+    if (e_array != NULL) {
+        stream_array_to_fd_bigset(e_array, &efds, &max_fd);
+    }
 
-	/* If seconds is not set to null, build the timeval, else we wait indefinitely */
-	if (!secnull) {
-		if (sec < 0) {
-			zend_argument_value_error(4, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		} else if (usec < 0) {
-			zend_argument_value_error(5, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		}
+    if (secnull && !usecnull) {
+        if (usec != 0) {
```

Comment:
This can probably be reduced to:

```c
if (secnull && !usecnull && usec != 0) {
```

Reply:
It was already like this in the old source. Don't know why. I updated the code with this change.